### ISSUE TITLE
Ray with docker: expose the jupyter UI

### ISSUE THREAD

**Issue by naifrec** [ray-project/ray#1919]
Hello all,

First, let me congratulate the whole team for this framework. This is **incredibly cool** and has empowered my research tremendously. It has a design philosophy that I love: only changes a few lines of the original sequential code and you are up and running. Now, onto my suggestion.

### System information
- Docker version 17.03.0-ce, build 60ccb22
- Ubuntu 16.04
- Ray 0.3.1
- Python 3.5.2

### Exposing Jupyter UI

I am using ray to parallelize training of RL agents across multiple nodes in a cluster. Each of these nodes are running ray inside a docker container. Now the head node is also running the UI under `localhost:8888`. However, even with the correct port forwarding this is not reachable from the outside world. I had to modify of `ray/services.py`, [in the `start_ui` function, line 773](https://github.com/ray-project/ray/blob/6f8b81d9e521a9a4aeb59478c0d1844f08d6ab81/python/ray/services.py#L737):

```python
    command = [
        "jupyter", "notebook", "--no-browser", "--port={}".format(port),
        "--ip=0.0.0.0",
        "--NotebookApp.iopub_data_rate_limit=10000000000",
        "--NotebookApp.open_browser=False",
        "--NotebookApp.token={}".format(token)
]
```

After this modification, I was able to reach the notebook. However, if I start the kernel and run all cells, nothing would ever show up, but nothing fails either, so I do not really know how to go from there.

In any case, the fact that nothing shows up might be related to my setup. However, my suggestion is to either default to `--ip=0.0.0.0` or let the use chose which IP to expose.

Cheers,
Guillaume

**Comment by robertnishihara** [comment]
Hi, I'm glad to hear that it's been helpful!

It might be worth checking the logs under `/tmp/raylogs/webui*` and see if some error message is printed by the UI.

Ignoring Ray, are you able to start a Jupyter notebook inside of a Docker container and then connect to it from the outside?

**Comment by naifrec** [comment]
Hello Robert,

- Ignoring Ray, are you able to start a Jupyter notebook inside of a Docker container and then connect to it from the outside?

Yes I am able to run a jupyter notebook inside a docker container and use it flawlessly. Within the docker container I run:

```bash
jupyter notebook --port=8888 --ip=0.0.0.0
```

And then I can access the notebook without any problem using the public hostname of the node and the port on my browser.

- It might be worth checking the logs under `/tmp/raylogs/webui*` and see if some error message is printed by the UI.

Thank you for the pointer, I looked into the `.err` log. It is a 6000 lines stack of the same error repeating itself:

```python-traceback
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tornado/gen.py", line 1107, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/local/lib/python3.5/dist-packages/tornado/websocket.py", line 810, in wrapper
    raise WebSocketClosedError()
tornado.websocket.WebSocketClosedError
ERROR:asyncio:Future exception was never retrieved
future: <Future finished exception=WebSocketClosedError()>
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tornado/websocket.py", line 808, in wrapper
    yield fut
  File "/usr/local/lib/python3.5/dist-packages/tornado/gen.py", line 1099, in run
    value = future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
tornado.iostream.StreamClosedError: Stream is closed

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tornado/gen.py", line 1107, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/local/lib/python3.5/dist-packages/tornado/websocket.py", line 810, in wrapper
    raise WebSocketClosedError()
tornado.websocket.WebSocketClosedError
ERROR:asyncio:Future exception was never retrieved
future: <Future finished exception=WebSocketClosedError()>
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/dist-packages/tornado/websocket.py", line 808, in wrapper
    yield fut
  File "/usr/local/lib/python3.5/dist-packages/tornado/gen.py", line 1099, in run
    value = future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 274, in result
    raise self._exception
tornado.iostream.StreamClosedError: Stream is closed

During handling of the above exception, another exception occurred:
```

Any idea on what is going on?

**Comment by robertnishihara** [comment]
Not sure about the tornado warning, though some other posts on the internet suggest it is innocuous.

Sounds like we should probably just use `--ip=0.0.0.0` (this is the same as `--ip=*`, right?). Any disadvantages to doing this?

Want to submit a PR with this change?

However, it'd be nice to figure out why nothing is showing up for you even when you make that change.

**Comment by RuofanKong** [comment]
I have the same issue - I could run a jupyter notebook inside the docker container (`jupyter notebook --port=8888 --no-browser --ip=0.0.0.0 --allow-root`) and get its data on my local browser (outside the container), but it's not working for ray jupyter notebook if I don't add `"--ip=0.0.0.0"` to the `command`.

**Comment by robertnishihara** [comment]
Thanks! @RuofanKong, would you be interested in submitting a PR fixing this?

**Comment by robertnishihara** [comment]
Fixed by #2883. Please let me know if the issue persists.
